### PR TITLE
Introduce "sort" in query to provide a general solution for sorting

### PR DESCRIPTION
**Pull request by ywk253100**
Introduce "sort" in query to provide a general solution for sorting

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: src/pkg/project/dao/dao.go [hunk 1]
```diff
@@ -94,13 +94,7 @@ func (d *dao) Create(ctx context.Context, project *models.Project) (int64, error
...
 // Count returns the total count of artifacts according to the query
 func (d *dao) Count(ctx context.Context, query *q.Query) (total int64, err error) {
-	query = q.MustClone(query)
-	query.Keywords["deleted"] = false
-	query.Sorting = ""
-	query.PageNumber = 0
-	query.PageSize = 0
-
-	qs, err := orm.QuerySetter(ctx, &models.Project{}, query)
+	qs, err := orm.QuerySetterForCount(ctx, &models.Project{}, query)
```
Comment (heww) [inline src/pkg/project/dao/dao.go:97]: `deleted=false` must in the keywords here.

File: src/pkg/blob/dao/dao.go [hunk 2]
```diff
@@ -222,10 +222,6 @@ func (d *dao) ListBlobs(ctx context.Context, query *q.Query) ([]*models.Blob, er
 	if err != nil {
 		return nil, err
 	}
-
-	if query.Sorting != "" {
-		qs = qs.OrderBy(query.Sorting)
-	}
 	blobs := []*models.Blob{}
```
Comment (heww) [inline src/pkg/blob/dao/dao.go:225]: The following in gc job should change to `Sorts` if `query.Sorting !=""` in this method was removed.
```golang
			blobRG := q.Range{
				Min: lastBlobID,
			}
			q := &q.Query{
				Keywords: map[string]interface{}{
					"update_time": &timeRG,
					"projectID":   p.ProjectID,
					"id":          &blobRG,
				},
				PageNumber: 1,
				PageSize:   int64(ps),
				Sorting:    "id",
			}
			blobs, err := gc.blobMgr.List(ctx.SystemContext(), q)
```

Comment (ywk253100) [reply src/pkg/blob/dao/dao.go:225]: done

File: src/jobservice/job/impl/gc/garbage_collection.go [hunk 3]
```diff
@@ -442,7 +442,11 @@ func (gc *GarbageCollector) removeUntaggedBlobs(ctx job.Context) {
 				},
 				PageNumber: 1,
 				PageSize:   int64(ps),
-				Sorting:    "id",
+				Sorts: []*q.Sort{
+					{
+						Key: "id",
+					},
+				},
```
Comment (heww) [inline src/jobservice/job/impl/gc/garbage_collection.go:449]: How about making the `parseSorting` in the q pkg be public and change these to `Sorts: q.ParseSorting("id")`?

Maybe I prefer `q.NewSorts("id")` than `q.ParseSorting("id")`.

File: src/core/api/search.go [hunk 4]
```diff
@@ -51,7 +51,11 @@ func (s *SearchAPI) Get() {
 	keyword := s.GetString("q")
 
 	query := q.New(q.KeyWords{})
-	query.Sorting = "name"
+	query.Sorts = []*q.Sort{
+		{
+			Key: "name",
+		},
+	}
```
Comment (heww) [inline src/core/api/search.go:58]: No need now because it's the default order by now.

File: src/pkg/repository/dao/dao.go [hunk 5]
```diff
@@ -71,7 +66,6 @@ func (d *dao) List(ctx context.Context, query *q.Query) ([]*models.RepoRecord, e
 	if err != nil {
 		return nil, err
 	}
-	qs = qs.OrderBy("-CreationTime", "RepositoryID")
```
Comment (heww) [inline src/pkg/repository/dao/dao.go:74]: What is the default order of the repositories when this line was removed?

File: src/pkg/tag/dao/dao.go [hunk 6]
```diff
@@ -71,7 +65,6 @@ func (d *dao) List(ctx context.Context, query *q.Query) ([]*tag.Tag, error) {
 	if err != nil {
 		return nil, err
 	}
-	qs = qs.OrderBy("-PushTime", "ID")
```
Comment (heww) [inline src/pkg/tag/dao/dao.go:74]: What is the default order for the tags when this line was removed?

**Review by heww — APPROVED** [review]

File: src/lib/orm/metadata.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+	// get the metadata from cache first
+	fullName := getFullName(t)
+	cacheMetadata, exist := cache.Load(fullName)
+	if exist {
+		return cacheMetadata.(*metadata)
+	}
+
+	// pointer value
+	v := reflect.ValueOf(model)
```
Comment (wy65701436) [inline src/lib/orm/metadata.go:73]: why not use the `ptr` here?

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/lib/orm/metadata.go (hunk 7 above)
Comment (ywk253100) [reply src/lib/orm/metadata.go:73]: `prt` is type, here is value

**Pull request merged**